Frequency Domain Chroma for X-Trans demosaicking

ILiebhardt opened the pull request:
This pull request introduces a new demosaicking option for X-Trans files, addressing Redmine Bug #10333, and comparable situations. Moiré is not completely avoided but substantially reduced. A new dependency is introduced: FFTW3. This dependency is optional, but compiling and running without FFTW3 will cause the demosaicking to run more slowly.

Contrary to closed pull request #1520, all commits are now squashed, clang style 3.6 is fully applied to demosaic.c via the beautify script, and initial observations from #1520 are addressed. Not addressed yet is the transitioning to modern cmake with imported targets, as this would be more a darktable-wide topic...

LebedevRI commented:
> Not addressed yet is the transitioning to modern cmake with imported targets, as this would be more a darktable-wide topic...

That was not a review note, just a generic observation.

ILiebhardt commented:
On the modern cmake as generic observation, you're right though. With all the dependencies, it would be a topic worth addressing ;-)

kubrickfr commented on src/iop/demosaic.c:
```diff
@@ -974,15 +977,1208 @@ static void xtrans_markesteijn_interpolate(float *out, const float *const in,
+#endif
+#ifndef HAVE_FFTW3
+            *(i_src + TS * (row - top) + (col - left)) = CLIP(in[roi_in->width * row + col]); // here is the
+                                                                                              // problem
```
@LebedevRI I guess the comments here refer to the green edge issue. Any case, these comments are not helpful without context and your workaround works well. Does the issue need to be investigated upstream though (rawspeed)?

LebedevRI replied:
Please provide the sample with the problem.

LebedevRI replied:
(it's impossible to say anything without seeing exactly that same sample where you have observed the issue.)

ILiebhardt replied:
It appears that I get values out of the expected range [0.f ... 1.f] back from 'in'. If you take the raw file [here](https://www.dropbox.com/s/o26gp98nwuli9qn/green_border_problem.RAF?dl=0) and process it with fdc demosaicking and remove the CLIPs, then you will see a green border on the right hand side. This is only for X-T2 images, X-Pro 1 images don't have this issue. 
The following code also remedies the issue:
`#ifdef HAVE_FFTW3
            *(i_src + TS * (row - top) + (col - left)) = ( (in[roi_in->width * row + col] < 0.f) ? 0.f : in[roi_in->width * row + col] )
                                                         + 0.f * _Complex_I;
#endif`
Hence my conclusion that _negative_ values must be the problem. I did the same check for greater than one, for isna() and for isinf(), without results. So negative values appear to be the reason.

LebedevRI replied:
I'm gonna bet the crop is wrong. Will check later.

ILiebhardt replied:
As to the question of upstream, I am not sufficiently the expert to know what processing sits between rawspeed itself and the data at this point, but I'll try to investigate further.

ILiebhardt replied:
Thanks @LebedevRI .

LebedevRI replied:
Yep, pretty clearly, i'm right, will adjust.

ILiebhardt replied:
Cool, thanks a lot. Then I'll remove the CLIPs from my side.

LebedevRI replied:
Pushed, hopefully that fixed the issue you were observing. Please remove that clipping.

ILiebhardt replied:
OK, I synced my fork back to upstream, taking your commit dc32f19 into account. I removed the CLIPping. Bad news: the issue isn't fully gone. Green border on the right hand side is still there, unfortunately.

ILiebhardt replied:
Hold on, my apologies. My bad. My eyes had been looking at too much code yesterday ;-)
`<Crop x="0" y="0" width="-132" height="0"/>`
did the trick! 1000times thanks.

kubrickfr commented on src/iop/demosaic.c:
```diff
@@ -1105,9 +2301,9 @@ static void lin_interpolate(float *out, const float *const in, const dt_iop_roi_
-                            const uint32_t filters, const uint8_t (*const xtrans)[6], const int only_vng_linear)
+static void vng_interpolate(float *out, const float *const in, const dt_iop_roi_t *const roi_out,
+                            const dt_iop_roi_t *const roi_in, const uint32_t filters,
+                            const uint8_t (*const xtrans)[6], const int only_vng_linear)
```
All these formatting changes should be remove from this pull request, don't you think? There's a lot of them and it makes the pull look unnecessarily big IMHO.

LebedevRI replied:
Heh, yeah. Never run clang-format over the whole file...
Use clang-format post-commit hook to format the diff.

kubrickfr commented on src/iop/demosaic.c:
```diff
@@ -3917,7 +5210,7 @@ static void demosaic_method_xtrans_callback(GtkWidget *combo, dt_iop_module_t *s
   if((p->demosaicing_method > (DT_IOP_DEMOSAIC_PASSTHROUGH_MONOCHROME | DEMOSAIC_XTRANS))
      || (p->demosaicing_method < DT_IOP_DEMOSAIC_VNG))
-    p->demosaicing_method = DT_IOP_DEMOSAIC_MARKESTEIJN;
+    p->demosaicing_method = DT_IOP_DEMOSAIC_FDC;
```
It needs to be discussed if FDC is to be made default or not as it changes the existing behaviour.

LebedevRI replied:
I posted that to the previous pr...

ILiebhardt replied:
To be honest, I would also still be conservative at this point and not make FDC the default yet. My impression is also that this must be clipping due to too big method id. I'll investigate and let you know.

ILiebhardt replied:
OK, I corrected this one, too. It wasn't strictly speaking the general default, it was the fallback for invalid values delivered back from the combobox. In any case, Markesteijn is now always default. On the way, I also spotted a minor error in the existing code: ' | DEMOSAIC_XTRANS' was missing once, I think.

ILiebhardt commented:
OK, to begin with, I applied the clang formatting to the differences selectively, instead of to the whole file. This should make the pull request much more readable again. I'll try to address the remaining topics one by one to take it further from here. @kubrickfr was and is a great help and contributor to me.

ILiebhardt commented:
OK, corrected the last one. Please let me know if you find something more. As to the mentioning of contributors (see also message from @kubrickfr), I'll leave it to the project maintainers, depending on what's preferred. I know that there's normally a 3 commits rule, but what was larger for this one isn't so much the commit itself but more the r&d effort that was behind. Unfortunately, with this exotic sensor, there is no conference paper saying: 'this is the perfect way, do it like this' ;-)

kubrickfr commented:
@ILiebhardt With FDC demozaicing, highlight recovery doesn't work very well.
With makesteijn: 
![image](https://user-images.githubusercontent.com/2634188/32146188-5fce1be0-bccb-11e7-803e-e9f5a72716ca.png)
With FDC (same settings):
![image](https://user-images.githubusercontent.com/2634188/32146192-6e9c7c16-bccb-11e7-80d9-4ea0d89fa3f7.png)

LebedevRI commented:
> With FDC demozaicing, highlight recovery doesn't work very well.

Probably because of the `CLIP()`. Is it *really* needed?

BTW i just feel like noting that we are in a feature freeze for `2.4.0` since `Oct 4`.

ILiebhardt commented:
Hi, I already removed the clipping in question with my commit from 2 September. 
As to the feature freeze, I had a PR in place before the closing date, so maybe it could still be taken up? Just let me know how you'd like to proceed.

LebedevRI commented on src/iop/demosaic.c:
```diff
@@ -985,6 +993,1201 @@ static void xtrans_markesteijn_interpolate(float *out, const float *const in,
+      for(int row = (top - sgrow + pad_rb_g + 2) / 3 * 3 + sgrow; row < mrow - pad_rb_g; row += 3)
+        for(int col = (left - sgcol + pad_rb_g + 2) / 3 * 3 + sgcol; col < mcol - pad_rb_g; col += 3)
+        {
+          float(*rfx)
```
Clearly, you did not.

LebedevRI commented on src/iop/demosaic.c:
```diff
@@ -985,6 +993,1201 @@ static void xtrans_markesteijn_interpolate(float *out, const float *const in,
+      for(int row = (top - sgrow + pad_rb_g + 2) / 3 * 3 + sgrow; row < mrow - pad_rb_g; row += 3)
+        for(int col = (left - sgcol + pad_rb_g + 2) / 3 * 3 + sgcol; col < mcol - pad_rb_g; col += 3)
+        {
+          float(*rfx)
```
Uhm, this does look like a clip, does it not?

ILiebhardt replied:
This is just part of the Markesteijn approach, which I merely took over. This is part of your codebase for a little while already, so I doubt that it makes the difference here. Actually, if you look at your initial review comments, you'll see that the clipping in question was at the input stage, and there, I removed it with my commit from 2 September.

LebedevRI replied:
github is showing the wrong diff, see screenshot.
![screenshot_20171029_212613](https://user-images.githubusercontent.com/88600/32146946-da9a176a-bcef-11e7-965c-b773dad3c3eb.png)

ILiebhardt replied:
Ah, thx for clarifying. 
But this is also not the clipping we discussed before, which was really at the input stage. (and that one, sorry, I removed it)
This clipping here is after chroma estimation. It's there to avoid values going out of range in areas of high color variations. As such, this is pretty much common practice in demosaicing and all other algorithms also have it in one form or the other. We could have a look how the demosaicing behaves removing it, but I remember that I put this one for a very specific reason.

ILiebhardt commented:
@kubrickfr : could you maybe provide me with the raw concerning the highlight recovery, such that I could try to investigate in more detail? Thx.

kubrickfr commented:
@ILiebhardt sure, https://drive.google.com/open?id=0ByPjAhK7ObclLVVrT2tQZ2hCeG8

ILiebhardt commented:
OK, I removed the clipping from the output stage entirely and I shifted the intermediary clipping to the chroma domain (it's only useful here). Results look much better now and highlight reconstruction should work properly.
@kubrickfr : thanks for the sample.
@LebedevRI : I'd be glad to have this taken up for 2.4.0 if still possible. This PR was in place well before the feature freeze; thus far I addressed all your review comments. Just let me know what you'd require from me. Thx.

houz commented on src/develop/develop.h:
```diff
@@ -150,6 +150,7 @@ typedef struct dt_develop_t
   // operations pipeline
   int32_t iop_instance;
   GList *iop;
+  dt_pthread_mutex_t fftw_lock;
```
What do you need the mutex for? Is libfftw3 not thread safe?

ILiebhardt replied:
Hi, 
The most relevant part of libfftw, the plan execution, is thread safe. But the plan creation and plan destruction are not. Performance wise, this is not a problem, because the plan execution is the really significant portion. If you'd like to dive more into the details, you could have a look in [the doc](http://fftw.org/fftw3.pdf) on page 51, subsection 5.4. Cheers, Ingo

ILiebhardt replied:
And to add, I only use the mutex for plan creation at the very beginning and for plan destruction & cleanup at the end.

houz replied:
I don't want to read the whole document now, but that section sounds as if the plan was something specific for the processed image at hand. Won't that break horribly when dt is processing two images in parallel, first creating the plan for image 1, releasing the lock, then creating the 2nd plan, then executing the plans. I expect both images working on the plan from the 2nd image. Or is the plan independent from the actual data and just some administrative data structures?

ILiebhardt replied:
It's the latter, only administrative data structures, depending exclusively on the tile size, which is constant. The plan is completely independent of the image content, so no worries in this respect. In theory it would even be possible to create a plan on darktable start-up and destroy it upon exit, but that would be at the expense of code readability in my opinion.

kubrickfr replied:
creating "the plan" on darktable start-up would break parallel processing, wouldn't it?

ILiebhardt replied:
@houz : thanks for the hint, I'll look into it! Why I was not too enthusiastic about it: in demosaic.c, we define the tile size locally as `#define TS 122`. If you look at the plan creation, the only real dependency is on this tile size, as I said above. Now, if we move the plan creation to `init_global()`, this is not so ideal. If you have a nice suggestion for me, I'm more than open.

ILiebhardt replied:
@kubrickfr : I don't think we would break parallel processing. In the present structure, the plan creation is in `static void xtrans_fdc_interpolate`, but before the actual OpenMP parallelism for the tiles starts. So now there is one single plan creation per demosaic, and then parallel use of the one plan for the multiple tiles. So I see little risk here...

ILiebhardt replied:
OK, I think we can live with alternatives concerning the TS definition. One important question for @houz is: is single-threaded execution of `init_global()` and `cleanup_global()` guaranteed? Otherwise, we're still confronted with the need for a mutex.

houz replied:
Yes, those are single threaded and called during dt initialization when loading the shared objects or closing dt.

kubrickfr replied:
From my experience, that's the reason why it was crashing : when zooming in and out,  the demozaic function was being called before the previous call finished/was cancelled , which is what was causing the crashes. If you move plan memory allocation outside of the function, maybe you'll avoid the crashes but you'll get some corruption / glitches I bet.

LebedevRI replied:
Seriously guys, you wanted us to break feature freeze and merge this code, despite that it apparently crashes during normal usage, *without* even mentioning that it  crashes during normal usage? As far as Ctrl+F can tell, this *is* the first mention.

kubrickfr replied:
@LebedevRI, yes indeed I meant  "crashing without that current lock" :) sorry if my comment was misleading.
I have been using this patch for months now with thousands of pictures, frequently rebased against the master branch and it has been very stable.
The point I was trying to make is that I think moving memory allocation outside the function would break things. Anyway, what's so problematic with having a mutex here?

LebedevRI replied:
So 
```diff
+// Preparations for fftw
+#ifdef HAVE_FFTW3
+  dt_pthread_mutex_lock(fftw_lock);
+  fftwf_complex *in_src = NULL, *out_src = NULL, *out_kernel = NULL, *in_kernel = NULL, *Cm = NULL;
+  // Initialization of the plans
+  fftwf_plan p_forw_src
+      = fftwf_plan_dft_2d(TS, TS, in_src, out_src, FFTW_FORWARD, FFTW_ESTIMATE); // real to complex
+  fftwf_plan p_forw_kernel = fftwf_plan_dft_2d(TS, TS, in_kernel, out_kernel, FFTW_FORWARD, FFTW_ESTIMATE);
+  fftwf_plan p_back = fftwf_plan_dft_2d(TS, TS, out_kernel, Cm, FFTW_BACKWARD, FFTW_ESTIMATE);
+  dt_pthread_mutex_unlock(fftw_lock);
+#endif
```
1. Does it actually depend on anything other than `TS` define?
2. Can you please replace that define with `static const int`.
3. Can the *exact same* plan be created once, and then used everywhere for everything, without creating it each time  in `xtrans_fdc_interpolate()`?

kubrickfr replied:
@LebedevRI 3) as I mentioned before when zooming in and out (for example) interpolation functions such as xtrans_fdc_interpolate() get called before the previous cancelled call is finished. I do not believe this is specific to xtrans_fdc_interpolate()

LebedevRI replied:
@kubrickfr that is orthogonal to my question

kubrickfr replied:
@LebedevRI Indeed, therefore no, I haven't tried, but I don't think you can uses the exact same plan created once

ILiebhardt replied:
@LebedevRI : to answer your three questions:
1. No, the only dependency is on TS, as I already said above.
2. Yes, I can make such replacement if you prefer that, no problem. I took over the `#define` from the existing code above.
3. I am pretty sure that the exact same plan be created once, and then used everywhere for everything, without creating it each time in `xtrans_fdc_interpolate()`. Already now, it is created once and used multiple times for all the tiles, and one roi can have hundreds of tiles. Busy day for me today, but I'll make the modifications tomorrow or on Friday and let you know.

ILiebhardt replied:
I removed the mutex and placed the plan creation and plan destruction in `init_global()` and `cleanup_global()`, respectively, as suggested by @houz . It doesn't crash on my machine, and following FFTW's documentation, I don't expect it to. I will follow this u with a rebase, and cleanup, as suggested by @LebedevRI . As said, I always addressed all your review comments thus far, and if you find something else, just let me know.

kubrickfr commented:
Yes, I remember putting this lock in place. If you remove it it crashes frequently.

aurelienpierre commented:
On a side note, @edgardoh and I will need FFTW3 too for the RL deconvolution module

LebedevRI commented:
Regardless, you probably want to rebase it and make sure than none of the commits changes e.g. `src/external/rawspeed`

>@aurelienpierre 
>On a side note, @edgardoh and I will need FFTW3 too for the RL deconvolution module

Not sure we like the trend of using some external libraries as hard dependencies for the whole modules.

ILiebhardt commented:
Hi @LebedevRI and @houz ,
I removed the mutex and the fftw plans are now in `init_global()` and `cleanup_global()`. It works fine, thanks again to @houz for the hint.
Furthermore, I rebased everything, cleaned up, and squashed all commits into a single one. I also made sure that none of the commits changes `src/external/rawspeed`. The only minor change that I was unable to do, unfortunately, is the tile size as `static const`. It is used in various places (also in the original Markesteijn) for defining array sizes and C doesn't allow for variable size arrays.

For everything else, all review comments were addressed.

If you find anything else, please just let me know.

ILiebhardt commented:
And one comment on the dependency on FFTW3: it is optional. There are code paths working without FFTW3 and delivering exactly the same image output. They are around 30% to 40% slower (depending on the machine), so FFTW3 is still highly recommended.

houz commented:
I found some time to look at the FFTW3 docs. To me it seems that reusing a plan is ONLY allowed with the "guru" interface and ONLY when array sizes, stride, ... don't change. That is obviously not the case for us so I am wondering if the lack of crashes and bad results is just luck so far. But maybe it just results in non-ideal runtimes. No idea. [0]
Also, the docs clearly say that the in and out arrays used for plan creation must be allocated, so NULL is not a valid argument. [1]

[0] http://www.fftw.org/fftw3_doc/New_002darray-Execute-Functions.html
[1] http://www.fftw.org/fftw3_doc/Complex-DFTs.html#Complex-DFTs

ILiebhardt commented:
Reusing a plan is not so much a matter of using the guru interface, but rather a matter of using the new array execute functions. This particular implementation hinges around the new array execute functions, they are key. I apply them to _tiles_, and due to this, array sizes, stride, ... don't change. The conditions of [0] are met.
As to [1], this is a bit more tricky. I found out in 'dry exercises' that NULL pointers in plan creation work if, and only if, you use the new array functions for execution, otherwise it fails. Now the question is: is the documentation a bit sparse on this matter (after all, the part on new array functions is a bit small) or is it working merely by coincidence?
We (including @kubrickfr) have done quite some testing, without crashes.
But if you find the use of FFTW3 too risky, I could also remove it completely and just go for the normal C implementation path without FFTW3 dependency. After all, the difference in speed is not that pronounced on stronger machines.
Just tell me if you prefer the latter and I'll quickly implement the changes.

LebedevRI commented:
> We (including @kubrickfr) have done quite some testing, without crashes.

Have you used ASAN?

ILiebhardt commented:
Me not, @kubrickfr used Valgrind.

ILiebhardt commented:
As said, the option without FFTW3 is absolutely viable for me, too.

houz commented:
>  I apply them to tiles, and due to this, array sizes, stride, ... don't change. The conditions of [0] are met.

Border tiles or small images can differ in size. The latter might not be a big problem for demosaic of course.

houz commented:
After discussing this a little more with @LebedevRI  and @hanatos we came to the conclusion that dropping FFTW3 is the best approach here. That way we can avoid a lot of potential bugs.

LebedevRI commented on src/iop/demosaic.c:
```diff
@@ -985,6 +998,1183 @@ static void xtrans_markesteijn_interpolate(float *out, const float *const in,
+  };
+
+
+  const float complex harr[4][13][13]
```
1. `static`, for all of these tables.
2. As discussed with @houz in IRC, surely, you did not come up with these tables by hand? There is an algorithm, right?

ILiebhardt replied:
1. Ok, no problem, I’ll drop FFTW3, much easier than adding it ;-)
2. I’ll make them static.
3. Sure there’s an algorithm. I sent @hanatos a rough overall description a while ago where it was also mentioned how these filters are made. It's basically training on reference images and then finding the least minimum square error. I implemented that in matlab, but it works also under gnu octave. Dirty code though, because it runs only once...
4. I can also give it a last check with asan, no problem. In any case dropping FFTW3 really reduces risks.

ILiebhardt replied:
I’ll do the above stuff tomorrow and let you know.

ILiebhardt commented:
Ah, and by the way for the tiles I used padding, so border tiles and small images are no problem.

kubrickfr commented:
For what it's worth, the implementation with FFTW3 has been extensively tested with thousands of my own pictures (and @ILiebhardt has a different camera model so I assume we have quite a bit of coverage). Assuming we drop it for now, which seems to be the way this is going, what needs to happen in order for it to be accepted in the future?

LebedevRI commented:
> Assuming we drop it for now, which seems to be the way this is going, what needs to happen in order for it to be accepted in the future?

It wasn't discussed as a temporary drop

LebedevRI commented:
>>> We (including @kubrickfr) have done quite some testing, without crashes.

>> Have you used ASAN?

> Me not, @kubrickfr used Valgrind.

It would seem i need to further clarify on the downvote.
The downvote is not because `@kubrickfr used Valgrind.`.
The downvote is because *it seems* you did not use anything.
Most importantly, the downvote is because you [both?] still didn't use AddressSanitizer, at the end of 2017.
There is pretty much no reason to not use it other than either `i don't know what that is` or `i'm maliciously/intentionally writing broken code`.
*Please* waste some of your precious time, and learn about it.
With any luck you will be scared shitless that you weren't using it before then...

Hope that clarified it...

(Note: since threading was discussed here, it would make sense to use threadsanitizer, but my previous expirience shows that getting it to work requires a lot of preparation.)

ILiebhardt commented:
OK, I entirely removed the FFTW3 part, changed the `const` occurrences to `static const`, as requested by @LebedevRI and squashed all into one commit.
I would now really like to give it a thorough check with asan. (for sure I'm not writing maliciously broken code, come on, I just would like to contribute)
I am not sure, though, if I'm checking properly. 

I built using `build.sh --asan`. 

Then I ran and expected some output on the shell, but nothing appeared.
I became sceptical and ran an `ldd darktable` and did not see lib asan explicitly.
I ran an `apt-get install libasan4`, but it's already installed.
`gcc --version` gives me 7.2.0.
`clang --version` gives me 4.0.1-6 
Independently of toolchain, any of the two should be new enough.

What could I be doing wrong? Or is it right and there's no memory bug?
Thanks for any help.

LebedevRI commented:
> I became sceptical and ran an ldd darktable and did not see lib asan explicitly.

It's linked statically by default, i'm not sure it will show up in `ldd` output.

> Then I ran and expected some output on the shell, but nothing appeared.

I'm not sure what this means, did darktable not start? If it started, then just use it as usual, if an error is detected, it will crash and you will see the output in console.

ILiebhardt commented:
Hmm, it started as usual.
It works as usual, albeit a little bit more slowly.
I used it a lot, particularly demosaicing with fdc, tiles, complete images, all...

No crash, no output.

So all appears to be fine.

Let me know if you still find something on your side. From my side, I think I addressed all.
Cheers, i

ILiebhardt commented:
Ah, hold on a sec, not really all, one last minor stupid thing, I forgot the conditional include of FFTW3 on top. Commit coming.

ILiebhardt commented:
Ok, now really all.

houz commented on src/iop/demosaic.c:
```diff
@@ -985,6 +990,1093 @@ static void xtrans_markesteijn_interpolate(float *out, const float *const in,
+  char *const all_buffers = (char *)dt_alloc_align(16, dt_get_num_threads() * buffer_size);
+  if(!all_buffers)
+  {
+    printf("[demosaic] not able to allocate FDC base buffers\n");
```
Please use `fprintf(stderr, ...)`

ILiebhardt replied:
OK, done and squashed into single commit.

hanatos commented:
hi, i just tried and merged this. it's... not sure it's a bugfix but whatever. i think the results really do look very good, and the patch is now not very intrusive at all any more. we can still improve speed in the future. fwiw, working with it in dr mode for a short while is noticeably slower than the default, but not absolutely unworkable, so i think people will find it very useful.

thanks for contributing this!

ILiebhardt commented:
hi hanatos, thx for merging, and my pleasure to contribute. I have less intrusive ideas for making it faster using sse, and I could also make an openCL version of it. That's for next year - after all it's good to have a nice project in the pipeline ;-)

dtorop commented:
A me-too note to say how exciting it is to so this show up in git master. I appreciate how nicely it handles fine detail without producing chroma noise. Much appreciation to all the work which went into its arriving.
